atlasdb-impl-shared/src/test/java/com/palantir/atlasdb/pue/CassandraImitatingConsensusForgettingStoreTest.java

```diff
@@ -0,0 +1,118 @@
+        neverThrowing.put(CELL, VALUE_2);
+        assertThat(neverThrowing.get(CELL).get()).hasValue(VALUE_2);
+    }
+
+    @Test
+    public void cannotPueTwice() {
+        neverThrowing.putUnlessExists(CELL, VALUE);
+        assertThatThrownBy(() -> neverThrowing.putUnlessExists(CELL, VALUE))
+                .isInstanceOf(KeyAlreadyExistsException.class)
+                .satisfies(exception -> assertThat(((KeyAlreadyExistsException) exception).getExistingKeys())
+                        .containsExactly(CELL));
+    }
+
+    @Test
+    public void canTouchAfterPue() throws ExecutionException, InterruptedException {
+        neverThrowing.putUnlessExists(CELL, VALUE);
+        neverThrowing.checkAndTouch(CELL, VALUE);
+        assertThat(neverThrowing.get(CELL).get()).hasValue(VALUE);
+    }
+
+    @Test
+    public void cannotTouchWhenNotMatching() {
+        assertThatThrownBy(() -> neverThrowing.checkAndTouch(CELL, VALUE))
+                .isInstanceOf(CheckAndSetException.class)
+                .satisfies(exception -> assertThat(((CheckAndSetException) exception).getActualValues())
+                        .isEmpty());
+
+        neverThrowing.putUnlessExists(CELL, VALUE);
+        assertThatThrownBy(() -> neverThrowing.checkAndTouch(CELL, VALUE_2))
+                .isInstanceOf(CheckAndSetException.class)
+                .satisfies(exception -> assertThat(((CheckAndSetException) exception).getActualValues())
+                        .containsExactly(VALUE));
+    }
+
+    @Test
+    public void testPartialFailures() throws ExecutionException, InterruptedException {
+        int numberOfSuccessfulPue = 0;
+        int numberOfNothingPresent = 0;
+        int numberOfValuePresentAfterFailure = 0;
+        for (int i = 0; i < 100; i++) {
+            Cell cell = Cell.create(PtBytes.toBytes(i), PtBytes.toBytes(i));
+            try {
+                sometimesThrowing.putUnlessExists(cell, VALUE);
+                numberOfSuccessfulPue++;
+                sometimesThrowing.setProbabilityOfFailure(0.0);
+                assertThat(sometimesThrowing.get(cell).get()).hasValue(VALUE);
+                sometimesThrowing.setProbabilityOfFailure(PROBABILITY_THROWING_ON_QUORUM_HALF);
+
+            } catch (RuntimeException e) {
+                sometimesThrowing.setProbabilityOfFailure(0.0);
+                Optional<byte[]> actualValue = sometimesThrowing.get(cell).get();
+                if (actualValue.isEmpty()) {
+                    numberOfNothingPresent++;
+                } else {
+                    assertThat(actualValue).hasValue(VALUE);
+                    numberOfValuePresentAfterFailure++;
+                }
+                sometimesThrowing.setProbabilityOfFailure(PROBABILITY_THROWING_ON_QUORUM_HALF);
+            }
+        }
+        assertThat(numberOfSuccessfulPue).isBetween(30, 70);
+        assertThat(numberOfNothingPresent).isBetween(5, 35);
+        assertThat(numberOfValuePresentAfterFailure).isBetween(15, 50);
```

Comment:
How did we get these numbers? I'm guessing through experimentation and/or solving the relevant Markov process, but they appear to be fairly permissive bounds.